export vector data in tif

ayat-khairy opened the pull request:
# PolarRoute Pull Request Template

Date: 31/5/2023
Version Number: 0.2.9
 
## Description of change
Handled the rendering of vector data (e.g. fuel/resistance) in the generated tif image. The current implementation calculates the mean of the vector data and displays in the mesh pixel. There were some infinity values in this vector data and they are rendered in red now to distinguish from the other values.


- [x] My changes result in all required regression tests passing without the need to update test files.  
[test_output.txt](https://github.com/antarctica/PolarRoute/files/11616542/test_output.txt)


# Checklist

- [x] My code follows [pep8](https://peps.python.org/pep-0008/) style guidelines.  
- [x] I have commented my code, particularly in hard-to-understand areas.  
- [x] I have updated the documentation of the codebase where required.  
- [x] My changes generate no new warnings.   
- [x] My PR has been made to the `0.2.x` branch (**DO NOT SUBMIT A PR TO MAIN**)

SamuelHall700 reviewed (changes requested):
The recent doc's changes in 0.2.x have yet to be pulled into this branch. Can they be pulled in first to avoid conflicts? 

The docs for for export_mesh CLI (`./docs/source/sections/Command_line_interface.rst`)will also need to be updated as they still refer to a 'color_conf' file which is no longer required.

ayat-khairy commented:
I have pulled the recent docs and merged.
I got rid of the colour_conf from the Doc when I switched to the greyscale default colour but then after facing the infinity issue. I switched to setting the colour back to blue-scale and set the infinity colour to red. So returned the colour_conf back to doc as optional parameter that might be set by the user. Happy to get rid of them though if needed.

SamuelHall700 reviewed (changes requested):
When attempting to export part of a mesh to .tif format using the CLI, the following error messages are given: 
```
(venv) C:\Users\samhall\source\repos\PolarRoute>export_mesh amsr_3dayhindcast_SDA.json TIF -o 01_06_23_fuel.tif -f configs/exports/fuel_tif_conf.json -v
 Mesh arg = <_io.TextIOWrapper name='amsr_3dayhindcast_SDA.json' mode='r' encoding='cp1252'>, format arg = TIF
[01-06-23 13:43:36 :INFO    ] - export_mesh 0.2.9
[01-06-23 13:43:36 :INFO    ] - exporting mesh to 01_06_23_fuel.tif in format TIF
[01-06-23 13:43:36 :INFO    ] - - saving the environment mesh to 01_06_23_fuel.tif
[01-06-23 13:43:36 :INFO    ] - Generating the tif image ...
ERROR 1: Deleting 01_06_23_fuel.tif failed:
Permission denied
ERROR 1: TIFFFetchDirectory:01_06_23_fuel.tif: Can not read TIFF directory count
ERROR 1: TIFFReadDirectory:Failed to read directory at offset 8
ERROR 1: _TIFFPartialReadStripArray:Cannot read offset/size for strile around ~0
ERROR 1: _TIFFPartialReadStripArray:Cannot read offset/size for strile around ~0
ERROR 1: 01_06_23_fuel.tif, band 1: IReadBlock failed at X offset 0, Y offset 0: _TIFFPartialReadStripArray:Cannot read offset/size for strile around ~0
ERROR 1: Deleting 01_06_23_fuel.tif failed:
Permission denied
0ERROR 1: _TIFFPartialReadStripArray:Cannot read offset/size for strile around ~0
ERROR 1: _TIFFPartialReadStripArray:Cannot read offset/size for strile around ~0
ERROR 1: 01_06_23_fuel.tif, band 1: IReadBlock failed at X offset 0, Y offset 0: _TIFFPartialReadStripArray:Cannot read offset/size for strile around ~0
[01-06-23 13:50:06 :INFO    ] - Generated GeoTIFF: 01_06_23_fuel.tif
[01-06-23 13:50:06 :INFO    ] - Timed call to export_mesh_cli took 390.168676 seconds
```
The functionality appears to work, but produces to files:
**01_06_23_fuel.tif**:
![image](https://github.com/antarctica/PolarRoute/assets/96304883/59c223f2-bb14-4ffe-92d0-600e23fe218e)
(appears to be empty)
**grid_data**:
![image](https://github.com/antarctica/PolarRoute/assets/96304883/b5e8fb39-04bb-4858-8ab3-d81350f8b1e8)
(appears to be the correct file, but has not been saved to the correct name)

The mesh and configs used for built it are attached here:
[configs.zip](https://github.com/antarctica/PolarRoute/files/11625726/configs.zip)


@ayat-khairy can you investigate?

ayat-khairy commented:
Good that it works on your end. I will investigate Sam's error. The 2 files are produced by GDAL to save and additional data, I will see if there is a way to get rid of them.

ayat-khairy commented:
@SamuelHall700 I can not reproduce the issue as I do not have any errors on my end as well. Here is the output I get
![image](https://github.com/antarctica/PolarRoute/assets/2110822/3593e9fd-e825-4735-b93f-4b55f3d7f1ab)

![image](https://github.com/antarctica/PolarRoute/assets/2110822/b3280216-1ae6-463c-832e-cb896daeeac4)

![image](https://github.com/antarctica/PolarRoute/assets/2110822/6108dcc4-ac15-4f6f-b0d5-8d69ba6fb996)

I think the grid data is the greyscale data in the default projection before transforming the projection and changing the colour. I can add few lines to the code to get rid of these additional files (grid_data, *.aux.xml)  if needed.

gecoombs commented:
I think it would be good to clean up those additional files unless there's an obvious potential use for them. If that were the case then we would want to document that and say why they're being saved. It would also be good to understand the errors Sam sees on Windows. The error message suggests it's something permissions related, have you checked if your user has write access in all the relevant directories? @SamuelHall700

ayat-khairy commented:
I have removed the unnecessary additional files generated during the tif generation.

gecoombs reviewed (approved):
I'm happy with this as it stands and all seems to work well on my system, I'll let @SamuelHall700 decide how to proceed with the issues he's seen on Windows.

SamuelHall700 commented:
Still having troubles getting this to work. I've attempted running with admin privileges but no luck. I can't seem to get it running on Windows or Linux either (the error message is slightly different though...)

```
(venv) (base) samhall@bs-samhall-l1:~/repos/PolarRoute$ export_mesh amsr_3dayhindcast_SDA.json TIF -f configs/exports/fuel_tif_conf.json -v -o SDA_fuel.tif
 Mesh arg = <_io.TextIOWrapper name='amsr_3dayhindcast_SDA.json' mode='r' encoding='UTF-8'>, format arg = TIF
[05-06-23 13:49:11 :INFO    ] - export_mesh 0.2.9
[05-06-23 13:49:12 :INFO    ] - exporting mesh to SDA_fuel.tif in format TIF
[05-06-23 13:49:12 :INFO    ] - - saving the environment mesh to SDA_fuel.tif
[05-06-23 13:49:12 :INFO    ] - Generating the tif image ...
0...10...20...30...40...50...60...70...80...90...100 - done.
free(): invalid pointer
Aborted
Traceback (most recent call last):
  File "/home/samhall/repos/PolarRoute/venv/bin/export_mesh", line 11, in <module>
    load_entry_point('polar-route', 'console_scripts', 'export_mesh')()
  File "/home/samhall/repos/PolarRoute/polar_route/utils.py", line 214, in wrapper
    res = func(*args, **kwargs)
  File "/home/samhall/repos/PolarRoute/polar_route/cli.py", line 181, in export_mesh_cli
    env_mesh.save(args.output, args.format , args.format_conf)
  File "/home/samhall/repos/PolarRoute/polar_route/mesh_generation/environment_mesh.py", line 441, in save
    self.to_tif(format_params, path)
  File "/home/samhall/repos/PolarRoute/polar_route/mesh_generation/environment_mesh.py", line 377, in to_tif
    set_colour(data, path, params)
  File "/home/samhall/repos/PolarRoute/polar_route/mesh_generation/environment_mesh.py", line 330, in set_colour
    subprocess.check_call(cmd, shell=True)
  File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'gdaldem color-relief SDA_fuel.tif /tmp/tmp627pkoai.txt SDA_fuel.tif' returned non-zero exit status 134.
double free or corruption (out)
Aborted
```
I'm gonna attempt some more debugging to see if I can figure this out by tomorrow

SamuelHall700 commented:
Despite the error messages, it looks life the .tif was successfully create in Ubuntu:
![image](https://github.com/antarctica/PolarRoute/assets/96304883/5fca19a9-89b4-47ef-9b79-7dd57ee9fb6f)

ayat-khairy commented:
Perhaps try running this command and see if GDAL works properly on your end  `gdaldem color-relief SDA_fuel.tif colour_conf.txt SDA_fuel.tif`

[colour_conf.txt](https://github.com/antarctica/PolarRoute/files/11654831/colour_conf.txt)

If this command works, then the issue would be related to create the tmp colour_conf file on your end.
